### PR TITLE
connection: adding watermarks to the read buffer.

### PR DESCRIPTION
**Pull request by alyssawilk**
Fixing an issue where every time a connection was readDisabled/readEnabled it would read from the socket, even if the buffer already contained sufficient data it should have triggered push back.

Risk Level: high (watermarks)
Testing: new unit tests
Docs Changes: n/a
Release Notes: n/a
Fixes https://github.com/envoyproxy/envoy-setec/issues/90 (no longer a security risk after previous mitigation)

### CODE REVIEW

File: test/common/network/connection_impl_test.cc [hunk 1]
```diff
@@ -791,6 +800,75 @@ TEST_P(ConnectionImplTest, Watermarks) {
...
+    EXPECT_FALSE(client_connection_->readEnabled());
+  }
+
+  // Now have the consumer read disable.
+  // This time when the buffer is drained, there will be no kick as the consumer
+  // does not want to read.
+  {
+    client_connection_->readDisable(true);
+    testClientConnection()->readBuffer().drain(3);
+    EXPECT_FALSE(testClientConnection()->readBuffer().aboveHighWatermark());
```
Comment (antoniovicente) [inline test/common/network/connection_impl_test.cc:862]: Worth adding some consumerWantsToRead and dispatch_buffered_data_ expectations?

File: source/common/network/connection_impl.cc [hunk 2]
```diff
@@ -465,9 +471,22 @@ void ConnectionImpl::setBufferLimits(uint32_t limit) {
...
+void ConnectionImpl::onReadBufferLowWatermark() {
+  ENVOY_CONN_LOG(debug, "onBelowReadBufferLowWatermark", *this);
+  if (state() == State::Open) {
+    readDisable(false);
+  }
+}
+
+void ConnectionImpl::onReadBufferHighWatermark() {
+  ENVOY_CONN_LOG(debug, "onAboveReadBufferHighWatermark", *this);
+  readDisable(true);
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply source/common/network/connection_impl.cc:487]: I'm happy to add a guard (you know how pro-error-handling I am) but I don't think it's necessary.
In the case we do SSL_read and pick up the fin, I believe we will do multiple reads, but in SslSocket::doRead we commit all the data (buffer goes high) and then return the action  PostIoAction::Close, so retaining the invariant that we close after the read is done, and shouldn't add data to the connection after the socket has been closed.

File: source/common/network/connection_impl.cc [hunk 3]
```diff
@@ -529,6 +548,19 @@ void ConnectionImpl::onReadReady() {
...
 
+  // We get here while read disabled in two ways.
+  // 1) There was a call to setReadBufferReady(), for example if a raw buffer socket ceded due to
+  //    shouldDrainReadBuffer(). In this case we defer the event until the socket is read enabled.
+  // 2) The consumer of connection data called readDisable(true), and instead of reading from the
+  //    socket we simply need to dispatch already read data.
+  if (read_disable_count_ != 0) {
+    if (dispatch_buffered_data_ && consumerWantsToRead()) {
+      onRead(read_buffer_.length());
+      dispatch_buffered_data_ = false;
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply source/common/network/connection_impl.cc:559]: Oh, that's a really nice catch.
Looking closely at the code, I was replicating the buggy logic from down below (583) where the same thing could happen on the actual socket read path.  I was going to bump both calls above the onRead, but then the lower case doesn't clear dispatch_buffered_data consistently, so I think I'm just going to latch-and-clear at the top of the function to attempt to future-proof.

File: source/common/network/connection_impl.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (antoniovicente) [reply source/common/network/connection_impl.cc]: I'm ok skipping the check.

File: source/common/network/connection_impl.h [hunk 4]
```diff
@@ -122,11 +122,19 @@ class ConnectionImpl : public ConnectionImplBase, public TransportSocketCallback
   static uint64_t nextGlobalIdForTest() { return next_global_id_; }
 
 protected:
+  // A convenience function which returns true if
+  // 1) The read disable count is zero or
+  // 2) The read disable count is one, due to the read buffer being overrun.
```
Comment (mattklein123) [inline source/common/network/connection_impl.h:127]: Isn't there a chance that the read disable count is > 1 and the buffer is overrun? Like if it got overrun but also disabled some other way? Or is the idea that if it is > 1 we will wait until the only reason is it's disabled due to read overrun? If that's the case can you add more comments just to make it completely clear?

File: source/common/network/connection_impl.cc [hunk 5]
```diff
@@ -342,24 +350,24 @@ void ConnectionImpl::readDisable(bool disable) {
     }
   } else {
     --read_disable_count_;
```
Comment (mattklein123) [inline source/common/network/connection_impl.cc:352]: While you are here can you ASSERT this is greater than 0 before decrementing? I'm surprised this was not already asserted.

**Pull request merged**